File: skills/skillcraft/scripts/quick_validate.py
#!/usr/bin/env python3
"""
Quick validation script for skills - minimal version
"""

import ast
import re
import sys
from pathlib import Path

import yaml

MAX_SKILL_NAME_LENGTH = 64


def _is_profile_expression(node):
    """Return whether an AST node reads a project profile identifier."""

    return (isinstance(node, ast.Name) and node.id == "profile") or (
        isinstance(node, ast.Attribute) and node.attr == "profile"
    )


def _contains_string_literal(node):
    """Return whether a comparison operand embeds a concrete string."""

    if isinstance(node, ast.Constant):
        return isinstance(node.value, str)
    if isinstance(node, (ast.List, ast.Set, ast.Tuple)):
        return any(_contains_string_literal(element) for element in node.elts)
    return False


def find_literal_profile_branch(skill_path):
    """Find reusable Python code branching on a concrete project profile."""

    scripts = Path(skill_path) / "scripts"
    if not scripts.is_dir():
        return None
    for path in sorted(scripts.rglob("*.py")):
        if "tests" in path.parts:
            continue
        try:
            tree = ast.parse(path.read_text(encoding="utf-8"), filename=str(path))
        except (OSError, SyntaxError):
            continue
        for node in ast.walk(tree):
            if isinstance(node, ast.Compare):
                operands = [node.left, *node.comparators]
                if any(_is_profile_expression(item) for item in operands) and any(
                    _contains_string_literal(item) for item in operands
                ):
                    return path, node.lineno
            if isinstance(node, ast.Match) and _is_profile_expression(node.subject):
                for case in node.cases:
                    if isinstance(case.pattern, ast.MatchValue) and _contains_string_literal(
                        case.pattern.value
                    ):
                        return path, case.pattern.lineno
    return None


def validate_skill(skill_path):
    """Basic validation of a skill"""
    skill_path = Path(skill_path)

    skill_md = skill_path / "SKILL.md"
    if not skill_md.exists():
        return False, "SKILL.md not found"

    content = skill_md.read_text()
    if not content.startswith("---"):
        return False, "No YAML frontmatter found"

    match = re.match(r"^---\n(.*?)\n---", content, re.DOTALL)
    if not match:
        return False, "Invalid frontmatter format"

    frontmatter_text = match.group(1)

    try:
        frontmatter = yaml.safe_load(frontmatter_text)
        if not isinstance(frontmatter, dict):
            return False, "Frontmatter must be a YAML dictionary"
    except yaml.YAMLError as e:
        return False, f"Invalid YAML in frontmatter: {e}"

    allowed_properties = {"name", "description", "license", "allowed-tools", "metadata"}

    unexpected_keys = set(frontmatter.keys()) - allowed_properties
    if unexpected_keys:
        allowed = ", ".join(sorted(allowed_properties))
        unexpected = ", ".join(sorted(unexpected_keys))
        return (
            False,
            f"Unexpected key(s) in SKILL.md frontmatter: {unexpected}. Allowed properties are: {allowed}",
        )

    if "name" not in frontmatter:
        return False, "Missing 'name' in frontmatter"
    if "description" not in frontmatter:
        return False, "Missing 'description' in frontmatter"

    name = frontmatter.get("name", "")
    if not isinstance(name, str):
        return False, f"Name must be a string, got {type(name).__name__}"
    name = name.strip()
    if name:
        if not re.match(r"^[a-z0-9-]+$", name):
            return (
                False,
                f"Name '{name}' should be hyphen-case (lowercase letters, digits, and hyphens only)",
            )
        if name.startswith("-") or name.endswith("-") or "--" in name:
            return (
                False,
                f"Name '{name}' cannot start/end with hyphen or contain consecutive hyphens",
            )
        if len(name) > MAX_SKILL_NAME_LENGTH:
            return (
                False,
                f"Name is too long ({len(name)} characters). "
                f"Maximum is {MAX_SKILL_NAME_LENGTH} characters.",
            )

    description = frontmatter.get("description", "")
    if not isinstance(description, str):
        return False, f"Description must be a string, got {type(description).__name__}"
    description = description.strip()
    if description:
        if "<" in description or ">" in description:
            return False, "Description cannot contain angle brackets (< or >)"
        if len(description) > 1024:
            return (
                False,
                f"Description is too long ({len(description)} characters). Maximum is 1024 characters.",
            )

    project_configured = (skill_path / "references/project_config.md").is_file()
    if project_configured:
        literal_branch = find_literal_profile_branch(skill_path)
        if literal_branch:
            path, line = literal_branch
            relative = path.relative_to(skill_path)
            return (
                False,
                f"{relative}:{line} branches on a concrete project profile. "
                "Treat profile names as opaque and move project behavior to "
                f".agents/skills-config/{name}/.",
            )

    return True, "Skill is valid!"


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python quick_validate.py <skill_directory>")
        sys.exit(1)

    valid, message = validate_skill(sys.argv[1])
    print(message)
    sys.exit(0 if valid else 1)

File: skills/skillcraft/scripts/quick_validate.mjs
#!/usr/bin/env node

import { readdirSync, readFileSync, statSync } from "node:fs";
import { join, relative, resolve } from "node:path";
import { pathToFileURL } from "node:url";
import { parseFrontmatter, FrontmatterError } from "./lib/frontmatter.mjs";

export const MAX_SKILL_NAME_LENGTH = 64;

export function findLiteralProfileBranch(skillPath) {
  const scripts = join(skillPath, "scripts");
  if (!isDirectory(scripts)) return null;
  for (const path of walkFiles(scripts).filter((item) => item.endsWith(".py")).sort()) {
    if (path.split("/").includes("tests")) continue;
    let source;
    try { source = readFileSync(path, "utf8"); } catch { continue; }
    const comparison = findProfileComparison(source);
    if (comparison) return { path, line: comparison.line };
    const matchCase = findProfileMatchCase(source);
    if (matchCase) return { path, line: matchCase.line };
  }
  return null;
}

export function validateSkill(skillPathInput) {
  const skillPath = resolve(skillPathInput);
  const skillMd = join(skillPath, "SKILL.md");
  if (!isFile(skillMd)) return [false, "SKILL.md not found"];

  const content = readFileSync(skillMd, "utf8");
  if (!content.startsWith("---")) return [false, "No YAML frontmatter found"];

  let frontmatter;
  try { frontmatter = parseFrontmatter(content); }
  catch (error) {
    const message = error instanceof FrontmatterError ? error.message : String(error);
    return [false, message];
  }

  const allowedProperties = new Set(["name", "description", "license", "allowed-tools", "metadata"]);
  const unexpected = Object.keys(frontmatter).filter((key) => !allowedProperties.has(key)).sort();
  if (unexpected.length > 0) {
    return [false, `Unexpected key(s) in SKILL.md frontmatter: ${unexpected.join(", ")}. Allowed properties are: ${[...allowedProperties].sort().join(", ")}`];
  }
  if (!("name" in frontmatter)) return [false, "Missing 'name' in frontmatter"];
  if (!("description" in frontmatter)) return [false, "Missing 'description' in frontmatter"];

  const name = frontmatter.name;
  if (typeof name !== "string") return [false, `Name must be a string, got ${typeName(name)}`];
  const normalizedName = name.trim();
  if (normalizedName) {
    if (!/^[a-z0-9-]+$/.test(normalizedName)) return [false, `Name '${normalizedName}' should be hyphen-case (lowercase letters, digits, and hyphens only)`];
    if (normalizedName.startsWith("-") || normalizedName.endsWith("-") || normalizedName.includes("--")) return [false, `Name '${normalizedName}' cannot start/end with hyphen or contain consecutive hyphens.`];
    if (normalizedName.length > MAX_SKILL_NAME_LENGTH) return [false, `Name is too long (${normalizedName.length} characters). Maximum is ${MAX_SKILL_NAME_LENGTH} characters.`];
  }

  const description = frontmatter.description;
  if (typeof description !== "string") return [false, `Description must be a string, got ${typeName(description)}`];
  const normalizedDescription = description.trim();
  if (normalizedDescription.includes("<") || normalizedDescription.includes(">")) return [false, "Description cannot contain angle brackets (< or >)"];
  if (normalizedDescription.length > 1024) return [false, `Description is too long (${normalizedDescription.length} characters). Maximum is 1024 characters.`];

  if (isFile(join(skillPath, "references", "project_config.md"))) {
    const literalBranch = findLiteralProfileBranch(skillPath);
    if (literalBranch) {
      return [false, `${relative(skillPath, literalBranch.path)}:${literalBranch.line} branches on a concrete project profile. Treat profile names as opaque and move project behavior to .agents/skills-config/${normalizedName}.`];
    }
  }
  return [true, "Skill is valid!"];
}

function findProfileComparison(source) {
  const profile = String.raw`(?:[A-Za-z_]\w*\.)*profile`;
  const string = String.raw`(?:"(?:\\.|[^"\\])*"|'(?:\\.|[^'\\])*')`;
  const literal = String.raw`(?:${string}|\[[^\]\n]*${string}[^\]\n]*\])`;
  const pattern = new RegExp(String.raw`${profile}\s*(?:==|!=|\b(?:not\s+in|in)\b)\s*${literal}|${literal}\s*(?:==|!=|\b(?:not\s+in|in)\b)\s*${profile}`);
  let offset = 0;
  for (const line of source.split(/\r?\n/)) {
    if (pattern.test(line)) return { line: offset + 1 };
    offset += 1;
  }
  return null;
}

function findProfileMatchCase(source) {
  const profile = String.raw`(?:[A-Za-z_]\w*\.)*profile`;
  const matchPattern = new RegExp(String.raw`^\s*match\s+${profile}\s*:`);
  const casePattern = /^\s*case\s+(?:"(?:\\.|[^"\\])*"|'(?:\\.|[^'\\])*')\s*:/;
  const lines = source.split(/\r?\n/);
  let inProfileMatch = false;
  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index];
    if (matchPattern.test(line)) { inProfileMatch = true; continue; }
    if (!inProfileMatch || !line.trim()) continue;
    if (casePattern.test(line)) return { line: index + 1 };
    if (!/^\s+/.test(line)) inProfileMatch = false;
  }
  return null;
}

function walkFiles(directory) {
  const result = [];
  for (const entry of readdirSync(directory, { withFileTypes: true })) {
    const path = join(directory, entry.name);
    if (entry.isDirectory()) result.push(...walkFiles(path));
    else if (entry.isFile()) result.push(path);
  }
  return result;
}

function isFile(path) { try { return statSync(path).isFile(); } catch { return false; } }
function isDirectory(path) { try { return statSync(path).isDirectory(); } catch { return false; } }
function typeName(value) { if (value === null) return "null"; if (Array.isArray(value)) return "array"; return typeof value; }

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  if (process.argv.length !== 3) {
    console.error("Usage: node scripts/quick_validate.mjs <skill_directory>");
    process.exitCode = 1;
  } else {
    const [valid, message] = validateSkill(process.argv[2]);
    console.log(message);
    process.exitCode = valid ? 0 : 1;
  }
}
